Lesson08-JS-V/homework/homework.js

```diff
@@ -0,0 +1,91 @@
+// Do not change any of the function names
+
+
+//PROBLEM 1 - VERSION 1
+//Error for both versions: createUserClass() › should return a user constructor that correctly builds user objects
+
+// create an ES6 class or a constructor function called User
+// it should accept an options object with username, name, email, and password properties
+// in the constructor set the username, name, email, and password properties
+// the constructor should have a method 'sayHi' on its prototype that returns the string 'Hello, my name is {{name}}'
+// {{name}} should be the name set on each instance
+// return the class
+function createUserClass() {
+    return User;
+}
+
+User.prototype.sayHi = function() {
+    return `Hello, my name is ${this.name}`;
+};
+
+
+function User (username, name, email, password) {
+  this.username = username;
+  this.name = name;
+  this.email = email;
+  this.password = password;
+}
+
+const userConstructor = createUserClass();
+const user = new userConstructor;
+user.sayHi();
+
+
+// PROBLEM 1 - VERSION 2 -- using class with same error about the constructor object 
+function createUserClass() {
+  return User;
+}
+
+class User {
+  constructor(username, name, email, password) {
+      this.username = username;
+      this.name = name;
+      this.email = email;
+      this.password = password;
+  }
+  sayHi(){
+      `Hello, my name is ${this.name}`;
+  }
+}
+
+
+//PROBLEM 2 --> this works
+// const userConstructor = createUserClass();
+// const user = new userConstructor();
+// user.sayHi();
+function addPrototypeMethod(Constructor) {
+    Constructor.prototype.sayHi = function(){
+        return 'Hello World!';
+    };
+}
+
+
+//PROBLEM 3 
+//Error: addReverseString(StringPrototype) › should add a reverse string method to the String prototype that returns a reversed version of the string
+
+// add a method to the string constructor's prototype that returns a reversed copy of the string
+// name this method reverse
+// hint:
+// you will need to use 'this' inside of reverse
+// function addReverseString() {
+// }
+
+function addReverseString() {
+    let listWord = [];
+    for (let i=this.string.length-1; i >=0; i--){
+        listWord.push(this.string[i]);
+    }
+    return listWord.join('');
+}
+
+String.prototype.reverse = function addReverseString(){};
```

Comment:
So I think you want to add the `reverse` method to the `addReverseString` function only if the latter is called. So you'll want to put line 81 inside the function above:

```
function addReverseString() {
  // add a method to the string constructor's prototype that returns a reversed copy of the string
  // name this method reverse
  // hint:
  // you will need to use 'this' inside of reverse
  String.prototype.reverse = function() {
    // you logic to reverse the string
  };
}
```

Reply:
@paulghaddad I updated my code and when I run it in my practice file I get the correct output: a reversed string. However, the test fails. It says: 

addReverseString(StringPrototype) › should add a reverse string method to the String prototype that returns a reversed version of the string
 TypeError: str.reverse is not a function

```
function addReverseString(string) {
    this.string = string;
 
    addReverseString.prototype.reverse = function() {
       let listWord = [];
       for (let i=this.string.length-   1; i >=0; i--){
          listWord.push(this.string[i]);
       }
       return listWord.join('');
    };  
 }
```
I didn't think you could (or should) define a prototype method inside the function. I thought prototype methods needed defined outside of the function (unless it's a class using the class keyword then you can add attributes and methods together).

`function addReverseString(string) ` doesn't appear to be a class-like function either bc it's not capitalized?